docs/proposals/201912_thanos_binary_index_header.md

```diff
@@ -0,0 +1,215 @@
+  * Get the label index and copy it to the local file.
+  * Open question: How to get [Posting size to fetch](./201912_thanos_binary_index_header.md#posting-size-to-fetch)?
+  
+With that effort building time and resources should be compared with downloading the prebuilt `index-header` from the bucket. This allows us to reduce complexity of the system as 
+we don't need to cache that in object storage by compactor or even sidecar.
+
+Thanks to this format we can reuse most of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664) code
+to load file. TODO: Benchmark if mmap + loading is actually faster than JSON parsing (3)
+
+This also proposes loading blocks on demand (4).
+
+This means maintaining pool with limited number of `index-header` files loaded at the time in Store Gateway.
+With an LRU logic we should be able to deduce which of blocks should be unloaded and left on the disk. 
+
+This proposes following algorithm:
+
+* blocks newer then X (e.g 2w) being always loaded
+* blocks older than X will be loaded on demand on query time and kept cached until evicted.
+* background eviction process unloads blocks without pending reads to the amount of Y (e.g 100 blocks) in LRU fashion. 
+ 
+Both X and Y being configurable.
+ 
+## Risks
+
+### Posting size to fetch
+
+While idea of combing different pieces of TSDB index as our index-header is great, unfortunately we heavily rely 
+on information about size of each posting represented as `postingRange.End`. 
+
+We need to know that apriori to know how to partition and how many bytes we need to fetch from the storage to get each posting: https://github.com/thanos-io/thanos/blob/7e11afe64af0c096743a3de8a594616abf52be45/pkg/store/bucket.go#L1567
+
+To calculate those sizes we use [`indexr.PostingsRanges()`](https://github.com/thanos-io/thanos/blob/7e11afe64af0c096743a3de8a594616abf52be45/pkg/block/index.go#L156-L155) which
+scans through `posting` section of the TSDB index. Having to fetch whole postings section just to get size of each posting 
+makes this proposal less valuable as we still need to download big part of index and traverse through it instead of what we propose in [#Proposal](./201912_thanos_binary_index_header.md#proposal)
+
+For series we don't know the exact size either, however we estimate max size of each series to be 64*1024. It depends on sane number of label pairs and chunks per series.
+We have really only one potential case when this was too low: https://github.com/thanos-io/thanos/issues/552. Decision about series this was made here: https://github.com/thanos-io/thanos/issues/146
+
+For postings it's more tricky as it depends on number of series in which given label pair exists. For worst case it can be even million label-pair for popular pairs like `__name__=http_request_http_duration_bucket` etc.
+
+We have few options:
+
+* Encode this posting size in TSDB index `PostingOffset`. 
```

Comment:
I'm not sure this is a real option. Even if such change will be accepted and merged (and I don't think), you would have to deal with all previously created blocks, which invalidates this option.

Reply:
No really. We need to pull PostingOffset anyway. (:

Reply:
I thought the option was to change the TSDB index to add the size/end to the postings offset table, but I guess I'm wrong. Can you elaborate on this option, please?

Reply:
Added all of options including those not choosen